scripts/stakingSimulator.js

```diff
@@ -0,0 +1,580 @@
+                validity = {isValid: true};
+                withdrawRuns = logResult(validity, withdrawRuns);
+                break;
+            case NO_ACTION:
+                console.log("do nothing for this epoch...");
+                // Advance time by a bit
+                let currentBlockTime = await Helper.getCurrentBlockTime();
+                await Helper.mineNewBlockAt(
+                    currentBlockTime + Helper.getRandomInt(10, epochPeriod.toString())
+                );
+                // validity = await verifyNoActionInvariants(kyberStaking, stakers, result);
+                validity = {isValid: true};
+                noActionRuns = logResult(validity, noActionRuns);
+                break;
+            default:
+                console.log("unexpected operation: " + operation);
+                break;
+        }
+    }
+
+    console.log(`--- SIM RESULTS ---`);
+    console.log(`Deposit: ${depositRuns}`);
+    console.log(`Delegate: ${delegateRuns}`);
+    console.log(`Withdraw: ${withdrawRuns}`);
+    console.log(`Do nothing: ${noActionRuns}`);
+}
+
+async function executeAndVerifyDepositInvariants(kyberStaking, result, epochPeriod) {
+    let isValid = true;
+    let initState = await getState(kyberStaking, result, null);
+    let currentBlockTime = await Helper.getCurrentBlockTime();
+    await Helper.setNextBlockTimestamp(currentBlockTime + Helper.getRandomInt(5, epochPeriod.toNumber() / 3));
+
+    // do deposit
+    if (result.isValid) {
+        try {
+            await kyberStaking.deposit(result.amount, {from: result.staker});
+        } catch(e) {
+            console.log('Valid deposit, but failed');
+            console.log(e);
+            return false;
+        }
+    } else {
+        await expectRevert.unspecified(
+            await kyberStaking.deposit(result.amount, {from: result.staker})
+        );
+    }
+
+    let newState = await getState(kyberStaking, result, initState.oldDelegateAddress);
+    isValid &= (await verifyDepositChanges(initState, newState, result));
+    return {
+        isValid: isValid,
+        states: {'initState': initState, 'newState': newState}
+    }
+}
+
+async function executeAndVerifyDelegateInvariants(kyberStaking, result, epochPeriod) {
+    let isValid = true;
+    let initState = await getState(kyberStaking, result, null);
+    let currentBlockTime = await Helper.getCurrentBlockTime();
+    await Helper.setNextBlockTimestamp(currentBlockTime + Helper.getRandomInt(5, epochPeriod.toNumber() / 3));
+
+    // do delegate
+    if (result.dAddress == zeroAddress) {
+        await expectRevert(
+            kyberStaking.delegate(result.dAddress, {from: result.staker}),
+            "delegate: delegated address 0"
+        );
+    } else {
+        await kyberStaking.delegate(result.dAddress, {from: result.staker});
+    }
+
+    let newState = await getState(kyberStaking, result, initState.oldDelegateAddress);
```

Comment:
`oldDelegateAddress` is computed after `delegate` call, if `delegate` call changes `oldDelegateAddress` wrongly we won't be able to notice right?

Reply:
`oldDelegateAddress` = `staker.getLatestStakerData.representative` before the delegate function is called